### DISCUSSION TITLE
@casl/mongoose usage with typegoose?

### DISCUSSION THREAD

**Discussion by johnsonjo4531** [stalniy/casl#549] [Q&A] [answered]
I'm trying to use the @casl/mongoose library with [typegoose](https://typegoose.github.io/typegoose/) any good way to go about this? I can get a reproduction up tomorrow sometime, but I'm ending my workday and just thought I would ask now rather than later on the of chance someone wants to be beyond helpful. This is what I've tried. 

```ts
export class Message extends NodeInterface {
  @prop()
  message: string
  ...
}

export const MessageModel = getModelForClass<typeof Message, QueryHelpers & AccessibleFieldsDocument>(Message, {
	schemaOptions: {
		timestamps: true
	}
});
```

If I end up getting this tomorrow without help I'll try and post what I did back here.

**Comment by Alex0007** [comment, marked as answer]
```typescript
import { AccessibleRecordModel } from "@casl/mongoose";
import { getModelForClass as getModelForClassOld } from "@typegoose/typegoose";
import { AnyParamConstructor, BeAnObject, IModelOptions } from "@typegoose/typegoose/lib/types";

export const getModelForClass = <
  U extends AnyParamConstructor<any>,
  QueryHelpers = BeAnObject
>(
  cl: U,
  options: IModelOptions
): AccessibleRecordModel<InstanceType<U>, QueryHelpers> & U => {
  return getModelForClassOld(cl, options) as any;
};

```

**Reply by danielbayerlein** [reply]
@Alex0007 Thank you for your suggestion. Your solution does not support instance methods. Can you confirm that?

> The 'this' context of type 'Document<unknown, any, Person> & Person & Required<{ _id: number; }>' is not assignable to method's 'this' of type 'DocumentType<Person, BeAnObject>'.
  Property 'typegooseName' is missing in type 'Document<unknown, any, Person> & Person & Required<{ _id: number; }>' but required in type 'IObjectWithTypegooseFunction'.

**Reply by Alex0007** [reply]
@danielbayerlein i updated code snippet, so it now supports instance methods

**Reply by danielbayerlein** [reply]
I can't see any change. The following code works for me:

```typescript
import type { AccessibleRecordModel } from '@casl/mongoose';
import { getModelForClass as getModelForClassOld } from '@typegoose/typegoose';
import type {
  AnyParamConstructor,
  BeAnObject,
  DocumentType,
  IModelOptions,
} from '@typegoose/typegoose/lib/types';

export const getModelForClass = <
  U extends AnyParamConstructor<any>,
  QueryHelpers = BeAnObject,
>(
  cl: U,
  options?: IModelOptions,
): AccessibleRecordModel<InstanceType<U> & DocumentType<U, QueryHelpers>, QueryHelpers> & U =>
  getModelForClassOld(cl, options) as any;
```

**Reply by johnsonjo4531** [reply]
Thank you both with your help on this. Here's a [working codesandbox](https://codesandbox.io/p/sandbox/mystifying-sky-fxkzb7?selection=%5B%7B%22endColumn%22%3A45%2C%22endLineNumber%22%3A59%2C%22startColumn%22%3A45%2C%22startLineNumber%22%3A59%7D%5D&file=%2Fsrc%2Findex.ts) for those who landed here and still may not know how to use typegoose :D

**Reply by johnsonjo4531** [reply]
This is what I'm using know BTWs I'll try to keep it up to date :D. I prefixed everything with My because it makes it easier to import the modified type with VS Code's quick fixes.

```ts
import typegoose from "@typegoose/typegoose";
import type {
	AnyParamConstructor,
	DocumentType,
	IModelOptions,
} from "@typegoose/typegoose/lib/types.js";
import type { AccessibleFieldsDocument, AccessibleModel } from "@casl/mongoose";
import { MyQueryHelpers } from "./QueryHelpers.js";

export const getMyModelForClass = <
	U extends AnyParamConstructor<unknown>,
	QueryHelpers = MyQueryHelpers
>(
	cl: U,
	options?: IModelOptions
): MyReturnModelType<U, QueryHelpers> =>
	typegoose.getModelForClass(cl, options) as any;

export type MyReturnModelType<
	U extends AnyParamConstructor<unknown> = AnyParamConstructor<unknown>,
	QueryHelpers = MyQueryHelpers
> = MyModelType<U, QueryHelpers> & U;

export type MyModelType<
	U extends AnyParamConstructor<unknown> = AnyParamConstructor<unknown>,
	QueryHelpers = MyQueryHelpers
> = AccessibleModel<
	MyDocumentType<InstanceType<U>, QueryHelpers>,
	QueryHelpers
>;
export type MyDocumentType<
	T = unknown,
	QueryHelpers = MyQueryHelpers
> = DocumentType<T, QueryHelpers> & AccessibleFieldsDocument;

```
